src/constructs/autoscaling/asg.ts

```diff
@@ -21,6 +32,50 @@ export interface GuAutoScalingGroupProps
   overrideId?: boolean;
 }
 
+/**
+ * `minimumCodeInstances` and `minimumProdInstances` determine the number of ec2 instances running
+ * in each stage under normal circumstances (i.e. when there are no deployment or scaling events in progress).
+ *
+ * `maximumCodeInstances` and `maximumProdInstances` are both optional. If omitted, these
+ * will be set to `minimumCodeInstances * 2` and `minimumProdInstances * 2`, respectively.
+ * This allows us to support Riff-Raff's autoscaling deployment type by default.
+ *
+ * The maximum capacity values can be set manually if you need to scale beyond the default limit (e.g. due to heavy traffic)
+ * or restrict scaling for a specific reason.
+ */
+export interface GuAsgCapacityProps {
+  minimumCodeInstances: number;
+  minimumProdInstances: number;
+  maximumCodeInstances?: number;
+  maximumProdInstances?: number;
+}
```

Comment:
This interface feels a bit verbose, I wonder if TS allows us to do something like this:

```typescript
interface ASGProps {
  min: number,
  max: number
}

interface GuASGProps {
  Stage.PROD: ASGProps
  [stage: Stage]?: ASGProps
}
```

Where usage then becomes:

```typescript
const myAsgProps = {
  Stage.PROD: {
    min: 3,
    max: 6
  },
  Stage.CODE: {
    min: 1,
    max: 2
  }
}
````

Reply:
Alas, I don't think it's possible to make `PROD` mandatory and other stages optional in `GuASGProps` as suggested above (nor do I think it's correct, as if we didn't define CODE values, but deployed to CODE, what do we do?).

Here's a [working example](https://www.typescriptlang.org/play?#code/KYOwrgtgBAygLgQwObCgbwFBSgYQPIAiAolALxQBE+xFANFlAAoBKhZlLhdGAvhhgEsQcYACcAZggDGqAIIwA4o1EB7AA4BndA2wQhALijgIAIzH1suhAA9Dxs6N784ATzWoFYeUtWb2mbABtDUQUKCFYUOAAXUNvZXUNAG5eFIwpFRAQqHEVFUNPeN8tcgDLTgJDND0QQwBmWigIG0MANh4LS2oiKprDAEZG5tsoACY+HiA) of what I'm talking about.

Reply:
The assumption with my implementation is that most people will omit the optional props anyway, giving them:

```
capacity: {
      minimumCodeInstances: 1,
      minimumProdInstances: 3,
    }
```

As we calculate sensible default max values here: https://github.com/guardian/cdk/pull/252/files#diff-b219fec113d015feaadba12ad710f805f4ba8fc1fcf3c270be34c8fd31b67c0fR66-R70

Reply:
What about something like this?

```typescript
enum Stage {
  CODE = "CODE",
  PROD = "PROD",
}

interface ASGProps {
  min: number;
  max: number;
}

type GuASGProps = Record<Stage, ASGProps>;

class GuASG {
  public static defaultProps: GuASGProps = {
    [Stage.PROD]: {
      min: 3,
      max: 6,
    },
    [Stage.CODE]: {
      min: 1,
      max: 2,
    },
  };

  constructor(props: GuASGProps = GuASG.defaultProps) {
    console.log(props);
  }
}

const vanillaASG = new GuASG();

const myCustomProps: GuASGProps = {
  ...GuASG.defaultProps,
  [Stage.PROD]: {
    min: 3,
    max: 6,
  },
};

const myCustomGuASG = new GuASG(myCustomProps);
```

My main concern is the verbosity of adding more mappings over time - I suspect we'd want to move the instance type to a mapping soon, for example. In the above proposal, this can be achieved by:

```typescript
interface ASGProps {
  min: number;
  max: number;
  instanceType: string;
}
```

As in, one extra line. The benefit here is a strong relation between the `Stage` enum and the ASG props.

(The downside of both is it becomes difficult to cater for CAPI's `zebra` stage!)

Reply:
Thanks for taking the time to code up this suggestion 💯  

I've tried to incorporate this (plus a few other small improvements) here: https://github.com/guardian/cdk/pull/252/commits/b3a4d9599875763995a986a60800d3e2132327b3. Please let me know if I've misunderstood anything!

> (The downside of both is it becomes difficult to cater for CAPI's zebra stage!)

Our [current `Stage` enum](https://github.com/guardian/cdk/blob/main/src/constants/stage.ts#L1-L7) already precludes this, but this PR certainly exacerbates the problem. 

I did have a go at implementing a version of this where an `AllowedStages` type could be passed in as a type parameter, but I got into a complete tangle and decided not to pursue this further yet! Hopefully we can cover the majority of stacks with `CODE` and `PROD` stages.

Reply:
> I did have a go at implementing a version of this where an `AllowedStages` type could be passed in as a type parameter, but I got into a complete tangle and decided not to pursue this further yet! 

Sounds promising! Happy to pair on this.